### PR TITLE
#3656 (1) Stock line page

### LINKED ISSUE #3656: Move `Stock line` modal to its own page
## The suggestion

Stock Line has had more functionality added to it and the modal is now overcrowded with tabs. 

It would be nice if clicking on a stock line will redirect the user to the stock line page instead of opening up a modal. This would also allow the repack modal to be refactored as a tab in the stock line window.

This can be done in two parts: 
1. Refactor the modal to its own page
2. Refactor `repacks` to be a tab in the `Stock line` page

### Why should we invest time in this?
Things will looks less crowded and makes everything more mobile friendly (especially when one wants to do a repack! see #3439)

### Are there any risks associated with this change?
Probably not, it's the same thing we do for `Items`, might have to re-train users but should be intuitive enough...

### How much effort is required?
Would need quite a bit of effort and maybe some UX input

### PR DESCRIPTION
**Pull request by CarlosNZ**
Fixes #3656 (Part 1 -- more to come)

# 👩🏻‍💻 What does this PR do?

- Moves the "view stock" modal to a stand-alone page, with tabs
- Refactor `useStockline` hook using simplified structure for get/create/update

## 💌 Any notes for the reviewer?

Sorry! It's already pretty massive, so figured it was best to leave "Inventory Adjustments" and "Repack" functionality out for now. They should be reasonably straightforward to migrate over (then again, I said that about this first part).

See inline comments for more specific details.

# 🧪 Testing

- [ ] Can click on a stock line in "View Stock" list and load the correct page
- [ ] Can click "New Stock" from list page and add a new stock line
- [ ] Can edit stock details on page, with "Cancel" and "Save" functionality behaving as expected
- [ ] Log and Ledger tabs display correct and function as before

### Known issues:

- the currency inputs don't update when "cancelling" the input as they don't respond to external change -- have made issue: #3959 
- have commented out the "plugin" functionality for now in "StockLineForm" as I'm not entirely clear how it's supposed to work -- will address it in next PR
- the Footer buttons are still a "little cosy" with the sides of the container. But this is a known issue (https://github.com/msupply-foundation/open-msupply/issues/3827) and affects all footers, so should be fixed generally not just for this case.

### CODE REVIEW

File: client/packages/system/src/Stock/Components/StockLineForm.tsx
Comment (CarlosNZ) [inline client/packages/system/src/Stock/Components/StockLineForm.tsx:81]: Do you think we need to debounce the input? We're only updating the Patch/Draft state, and it seems fine to update this on each keystroke (not really much different to updating a `useState` within a component). What do we think?

File: client/packages/system/src/Stock/Components/StockLineForm.tsx
Comment (CarlosNZ) [inline client/packages/system/src/Stock/Components/StockLineForm.tsx:120]: This component is displayed in both a modal (when creating new) and on the page, so requires different spacing depending on which it is.

File: client/packages/system/src/Stock/Components/StockLineForm.tsx
Comment (CarlosNZ) [inline client/packages/system/src/Stock/Components/StockLineForm.tsx:123]: This stuff hasn't really changed, probably want to disable white-space changes when reviewing this.

File: client/packages/system/src/Stock/Components/StockLineForm.tsx
Comment (CarlosNZ) [inline client/packages/system/src/Stock/Components/StockLineForm.tsx:282]: Normally the Footer is rendered in with the DetailView component, but here we only want the Footer buttons on the main Details tab, so rendering it here instead. It's conditional, as we don't display it when this form appears in Modal.

File: client/packages/system/src/Stock/Components/StockLineForm.tsx
Comment (CarlosNZ) [reply client/packages/system/src/Stock/Components/StockLineForm.tsx:120]: Are we happy with the Modal for "Create new", or should we launch straight into an empty page? (I think this modal is fine, personally)

File: client/packages/system/src/Stock/DetailView/AppBarButtons.tsx
Comment (CarlosNZ) [inline client/packages/system/src/Stock/DetailView/AppBarButtons.tsx:32]: Not sure about this icon. There was nothing in the icon library that seemed totally appropriate -- can add a new one if we think it needs it?

File: client/packages/system/src/Stock/api/hooks/useStockLine.ts
Comment (CarlosNZ) [inline client/packages/system/src/Stock/api/hooks/useStockLine.ts:40]: We can use the same "patch" object for both "Create" and "Update" operations. :)

File: client/packages/system/src/Stock/api/hooks/useStockLine.ts
Comment (CarlosNZ) [inline client/packages/system/src/Stock/api/hooks/useStockLine.ts:57]: Now draft is just a derived value -- from merging any changes values (in patch) into the original database data.

File: client/packages/system/src/Stock/api/hooks/useStockLine.ts
Comment (CarlosNZ) [inline client/packages/system/src/Stock/api/hooks/useStockLine.ts:74]: "resetDraft" used when Cancelling an edit .

File: client/packages/system/src/Stock/api/hooks/useStockLine.ts
Comment (CarlosNZ) [inline client/packages/system/src/Stock/api/hooks/useStockLine.ts:87]: Don't love this duplicated nested naming -- any better suggestions?

File: client/packages/system/src/Stock/DetailView/DetailView.tsx
Comment (CarlosNZ) [inline client/packages/system/src/Stock/DetailView/DetailView.tsx:62]: Not sure exactly what to display for error here.

File: client/packages/system/src/Stock/DetailView/DetailView.tsx
Comment (CarlosNZ) [inline client/packages/system/src/Stock/DetailView/DetailView.tsx:125]: I don't think we're going to want any "Toolbar" functionality in here are we?

File: client/packages/system/src/Stock/DetailView/AppBarButtons.tsx
Comment (lache-melvin) [reply client/packages/system/src/Stock/DetailView/AppBarButtons.tsx:32]: Oof yeah that one looks weird... mayybe RefreshIcon? Though [this](https://mui.com/material-ui/material-icons/?query=arrow+up+down&selected=SwapVert) might be nicer?

File: client/packages/system/src/Stock/ListView/ListView.tsx
Comment (lache-melvin) [inline client/packages/system/src/Stock/ListView/ListView.tsx:24]:
```suggestion
import { AppRoute } from '@openmsupply-client/config';
```
Otherwise tests break 🥲

File: client/packages/system/src/Stock/ListView/ListView.tsx
Comment (lache-melvin) [inline client/packages/system/src/Stock/ListView/ListView.tsx:61]: I think a lot of this commented stuff could be removed?

File: client/packages/system/src/Stock/Components/Ledger/LedgerForm.tsx
Comment (lache-melvin) [inline client/packages/system/src/Stock/Components/Ledger/LedgerForm.tsx:29]: Not related to the changes, but why is this component called LedgerForm? Just display only right?

File: client/packages/system/src/Stock/Components/NewStockLineModal.tsx
Comment (lache-melvin) [inline client/packages/system/src/Stock/Components/NewStockLineModal.tsx:40]: Sorry can't put comment on the `save` method below, but wondering if a redirect to the detail page would be nicer than just the toast popup on save?

Maybe worth asking for UX input, but given we don't want `New Stock` to be used often/to add lots of stock lines in one go, I think it's nice to be redirected so you can see your saved changes?

Found myself searching through the stock list trying to find the stock line I just created, hoping it worked 😅

File: client/packages/system/src/Stock/DetailView/DetailView.tsx
Comment (lache-melvin) [reply client/packages/system/src/Stock/DetailView/DetailView.tsx:120]: Not at this stage I don't think so?

File: client/packages/system/src/Stock/Components/StockLineForm.tsx
Comment (lache-melvin) [reply client/packages/system/src/Stock/Components/StockLineForm.tsx:282]: I think it would be better to leave the footer in the DetailView component, and just keep the buttons disabled on the other tabs?

I don't think hiding it from view on the other tabs adds much value, and conditionally rendering a footer from a reusable form component definitely seems wrong 😅

File: client/packages/system/src/Stock/Components/StockLineForm.tsx
Comment (lache-melvin) [reply client/packages/system/src/Stock/Components/StockLineForm.tsx:81]: Seems fine to me 🤷‍♀️ I would have thought we would only need debounce if the onUpdate made an API call, or maybe if it resulted in a big table rerender

File: client/packages/system/src/Stock/DetailView/Footer.tsx
Comment (lache-melvin) [inline client/packages/system/src/Stock/DetailView/Footer.tsx:52]: Can this be:

```suggestion
              <DialogButton
                variant="cancel"
                disabled={!isDirty || isSaving}
                onClick={() => showCancelConfirmation()}
              />
```
The save button disabling/enabling, but the cancel button appearing/disappearing is weird. Usual design has been disable everywhere rather than hide?

File: client/packages/system/src/Stock/DetailView/DetailView.tsx
Comment (lache-melvin) [inline client/packages/system/src/Stock/DetailView/DetailView.tsx:33]: Was this a previous edition of `isDirty`?

File: client/packages/system/src/Stock/DetailView/DetailView.tsx
Comment (lache-melvin) [reply client/packages/system/src/Stock/DetailView/DetailView.tsx:58]: Do you also get two identical error snacks from this e.g. if you stop the server so network request fails? I think `GqlContext` might catch and show error snack for standard graphql errors... but we'd still need a catch block here for anything that could error before the API call, or mapping after? Not sure what the best way to do that is tbh... wonder if @mark-prins has any tips?

File: client/packages/system/src/Stock/api/hooks/useStockLine.ts
Comment (lache-melvin) [inline client/packages/system/src/Stock/api/hooks/useStockLine.ts:19]:
```suggestion
const defaultDraftStockLine: DraftStockLine = {
```
nice to have type safety in the future

File: client/packages/system/src/Stock/api/hooks/useStockLine.ts
Comment (lache-melvin) [inline client/packages/system/src/Stock/api/hooks/useStockLine.ts:64]: Was this needed because of calling updatePatch with the full object somewhere? Would have thought just updating the patch with the `newData` would have been fine??

File: client/packages/system/src/Stock/api/hooks/useStockLine.ts
Comment (lache-melvin) [reply client/packages/system/src/Stock/api/hooks/useStockLine.ts:87]: Don't really know if it's better, but an idea:
```ts
  data: data?.nodes[0],
  create,
  update,
  loading: { create: loadingCreate, update: loadingUpdate, data: loadingQuery, any: loadingCreate || loadingUpdate || loadingQuery},
  error: { create: createError, update: updateError, data: queryError },
```

File: client/packages/system/src/Stock/Components/StockLineForm.tsx
Comment (lache-melvin) [reply client/packages/system/src/Stock/Components/StockLineForm.tsx:120]: I think the modal is fine, its pretty consistent with the create patterns we have elsewhere. Might defer to @mark-prins on that though?

**Review by lache-melvin — CHANGES_REQUESTED** [review]
Looks awesome, thanks Carl! I left a lot of comments but most aren't big 👍 

Main concern is moving the footer back to the detail view, and the package/src import 😎

File: client/packages/system/src/Stock/Components/StockLineForm.tsx
Comment (CarlosNZ) [reply client/packages/system/src/Stock/Components/StockLineForm.tsx:282]: Okay, have moved back, but hidden on other tabs -- seems okay to me.

File: client/packages/system/src/Stock/api/hooks/useStockLine.ts
Comment (CarlosNZ) [reply client/packages/system/src/Stock/api/hooks/useStockLine.ts:87]: Seems about the same degree of confusion to me 😆

Might leave it for now, see if the wider group has opinions...

File: client/packages/system/src/Stock/ListView/ListView.tsx
Comment (CarlosNZ) [reply client/packages/system/src/Stock/ListView/ListView.tsx:61]: Yeah, have got rid of it on the next PR. Just left it here for now in case I wanted to use some it (which I have). I won't merge this one until I've got them both ready anyway.

File: client/packages/system/src/Stock/Components/Ledger/LedgerForm.tsx
Comment (CarlosNZ) [reply client/packages/system/src/Stock/Components/Ledger/LedgerForm.tsx:29]: Hm, no idea. Changed it to LedgerTable :)

File: client/packages/system/src/Stock/Components/NewStockLineModal.tsx
Comment (CarlosNZ) [reply client/packages/system/src/Stock/Components/NewStockLineModal.tsx:44]: > Found myself searching through the stock list trying to find the stock line I just created, hoping it worked 😅

Me too!

Good idea: d64e534d4

File: client/packages/common/src/ui/icons/BarChart2.tsx
Comment (lache-melvin) [inline client/packages/common/src/ui/icons/BarChart2.tsx:4]: Should it be called `TrendingUp` throughout?

File: client/packages/system/src/Stock/ListView/ListView.tsx
Comment (lache-melvin) [reply client/packages/system/src/Stock/ListView/ListView.tsx:61]: Awesome thanks 🙏

File: client/packages/system/src/Stock/DetailView/DetailView.tsx
Comment (CarlosNZ) [reply client/packages/system/src/Stock/DetailView/DetailView.tsx:35]: Yeah I guess it was. Sorry, I actually assumed it was something to do with the plugins, but at closer inspection it's not -- gone now 💨

File: client/packages/system/src/Stock/api/hooks/useStockLine.ts
Comment (CarlosNZ) [reply client/packages/system/src/Stock/api/hooks/useStockLine.ts]: 79aae858d

File: client/packages/system/src/Stock/api/hooks/useStockLine.ts
Comment (CarlosNZ) [reply client/packages/system/src/Stock/api/hooks/useStockLine.ts:64]: It prevents that same problem you dealt with the other day -- clicking and leaving an input triggers `onChange`, so you get the buttons becoming enabled even when you haven't actually changed anything. 

So we just check for actual changed values and filter out the ones that aren't.

File: client/packages/common/src/ui/icons/BarChart2.tsx
Comment (CarlosNZ) [reply client/packages/common/src/ui/icons/BarChart2.tsx:4]: I changed the icon, due to those weird prop/styling issues I mentioned in the chat. This one is called BarChart2: https://feathericons.com/?query=barchart-2

File: client/packages/common/src/ui/icons/index.ts
Comment (lache-melvin) [inline client/packages/common/src/ui/icons/index.ts:68]: > I changed the icon, due to those weird prop/styling issues I mentioned in the chat. This one is called BarChart2: https://feathericons.com/?query=barchart-2

But you're renaming here?

Comment (CarlosNZ) [reply client/packages/common/src/ui/icons/index.ts]: Oh, sorry I missed that. Fixed now: 393a89062

File: client/packages/system/src/Stock/Components/Ledger/LedgerForm.tsx
Comment (lache-melvin) [reply client/packages/system/src/Stock/Components/Ledger/LedgerForm.tsx:29]: Thanks 🙏 Probably also worth renaming the file?

**Review by lache-melvin — APPROVED** [review]
Meean thanks Carl, looking great!

File: client/packages/system/src/Stock/api/hooks/useStockLine.ts
Comment (lache-melvin) [reply client/packages/system/src/Stock/api/hooks/useStockLine.ts:64]: Hi - I think this issue should be solved in the `CurrencyInput` component (can be another issue/PR), or at least in the `StockLineForm` rather than adding more logic to these API hooks to compensate... should just be able to do `setPatch({ ...patch, newData })` here?

File: client/packages/system/src/Stock/api/hooks/useStockLine.ts
Comment (lache-melvin) [reply client/packages/system/src/Stock/api/hooks/useStockLine.ts:57]:
```suggestion
  const draft = { ...defaultDraftStockLine, ...data?.nodes[0], ...patch };
```

Sorry missed this before - we shouldn't have `as` anywhere if we can help it (I think you're needing it here because data.nodes[0] could be undefined, and also because the queried stock line doesn't have an inventoryAdjustmentReasonId, which the draft requires...)

**Review by lache-melvin — CHANGES_REQUESTED** [review]

File: client/packages/system/src/Stock/api/hooks/useStockLine.ts
Comment (CarlosNZ) [reply client/packages/system/src/Stock/api/hooks/useStockLine.ts]: Okay, have changed it a bit, so now it'll compare the current patch state with the saved data and toggle `isDirty` accordingly (if actually different):



f4e1bf21b

**Review by lache-melvin — APPROVED** [review]
That isDirty improvement looks awesome! Thanks 🚀

**Pull request merged**
